### PR TITLE
AdaptiveDynaMOSA implementation

### PR DESCRIPTION
**Pull request by giograno**
Hi all, this is the implementation of Adaptive DynaMOSA from the [paper](https://ieeexplore.ieee.org/document/8865437) "_Testing with Fewer Resources: An Adaptive Approach to Performance-Aware Test Case Generation_".

Probably @apanichella is one of the best candidates to look a it :)

### CODE REVIEW

**Comment by apanichella** [comment]
I'll give a look over the weekend (and make some local tests). I'll keep you posted.

@giograno can you add some test cases?

**Comment by giograno** [comment]
Thanks @apanichella! 

In the meanwhile, I added a few tests, especially for the sorting of the performance score and the indicators. 
Let me know if you need something else and/or you have any question.

**Comment by apanichella** [comment]
Great! I am gonna run some experiments to be sure we don't break the latest changes in the main/master branch

**Comment by giograno** [comment]
Hi there! Is there any news about this PR? Is still under consideration? Thanks :)

File: client/src/main/java/org/evosuite/ga/metaheuristics/mosa/AdaptiveDynaMOSA.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+        }
+
+        this.currentIteration++;
+//        logger.debug("N. fronts = {}", this.rankingFunction.getNumberOfSubfronts());
+//        logger.debug("1* front size = {}", this.rankingFunction.getSubfront(0).size());
+//        logger.debug("Best Values Map size = {}", goalsManager.getBestValues().size());
+        logger.debug("Covered goals = {}", goalsManager.getCoveredGoals().size());
+        logger.debug("Current goals = {}", goalsManager.getCurrentGoals().size());
+        logger.debug("Uncovered goals = {}", goalsManager.getUncoveredGoals().size());
```
Comment (shabbirabdul) [inline client/src/main/java/org/evosuite/ga/metaheuristics/mosa/AdaptiveDynaMOSA.java:135]: Can be condensed to 

`
logger.debug("Covered goals = {}, Current goals = {} and Uncovered goals = {}", goalsManager.getCoveredGoals().size(),   goalsManager.getCurrentGoals().size(), goalsManager.getUncoveredGoals().size())
`

File: client/src/main/java/org/evosuite/ga/metaheuristics/mosa/AdaptiveDynaMOSA.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+            for (int k = 0; k < remain; k++)
+                this.population.add(front.get(k));
+
+        }
+
+        this.currentIteration++;
+//        logger.debug("N. fronts = {}", this.rankingFunction.getNumberOfSubfronts());
+//        logger.debug("1* front size = {}", this.rankingFunction.getSubfront(0).size());
+//        logger.debug("Best Values Map size = {}", goalsManager.getBestValues().size());
```
Comment (shabbirabdul) [inline client/src/main/java/org/evosuite/ga/metaheuristics/mosa/AdaptiveDynaMOSA.java:132]: Commented code is harmless, however commented code greatly reduces the readability of code. I suggest we remove these commented lines.

**Comment by giograno** [comment]
Hi @shabbirabdul, many thanks for your comments. I have addressed them! Plus, I rebased the implementation of AdaptiveDynaMOSA on the master, since a lot has changed from when I opened the PR.

**Comment by giograno** [comment]
Hi there, I stumble across this PR, close to have its first birthday :) @apanichella do you have maybe any update? Many thanks 🙏

**Comment by giograno** [comment]
Hey there. This PR grew quite old and it never seemed to be under consideration. I am going to close it. 
Please, feel free to re-open it in case you are interested to the contribution. Thank you very much.

**Pull request closed without merging**